integration-tests/tests/src/tests/notifications.ts

```diff
@@ -0,0 +1,463 @@
+      });
+
+      this.realm.removeAllListeners("change");
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 2 });
+      });
+
+      expect(runCount).equal(2);
+    });
+  });
+
+  /**
+   * Helper function to use as a generic handler for collection listeners.
+   * @param onEnd Callback to run once the listener reaches the end of the changes array. If the callback is Mocha.Done, it will ensure that
+   * the listener was run exactly [changesOnRun.length] times
+   * @param changesOnRun An array of Realm.CollectionChangeSet.
+   * The handler will assert equality of listener's changes with the next element of the array on every run.
+   */
+  const expectCollectionChangesOnEveryRun = (
+    onEnd: () => void | Mocha.Done,
+    changesOnRun: Realm.CollectionChangeSet[],
+  ) => {
+    return (_: Realm.Collection<Realm.Object>, changes: Realm.CollectionChangeSet) => {
+      runCount++;
+
+      expect(changes.insertions).deep.equals(changesOnRun[runCount - 1].insertions);
+      expect(changes.oldModifications).deep.equals(changesOnRun[runCount - 1].oldModifications);
+      expect(changes.newModifications).deep.equals(changesOnRun[runCount - 1].newModifications);
+      // expect(changes.modifications).deep.equals(changesOnRun[runCount - 1].modifications);
+      expect(changes.deletions).deep.equals(changesOnRun[runCount - 1].deletions);
+
+      if (runCount >= changesOnRun.length) {
+        // Once runCount reaches given array, run the onEnd callback. If onEnd is Mocha.done then
+        // running it multiple times will cause tests to fail so this also ensures the listener fires
+        // exactly changesOnRun.length times.
+        onEnd();
+      }
+    };
+  };
+
+  describe("Results notifications", () => {
+    let testObjects: Realm.Collection<TestObject>;
+
+    openRealmBeforeEach({ schema: [TestObject] });
+
+    beforeEach(function (this: RealmContext) {
+      testObjects = this.realm.objects(TestObject);
+    });
+
+    it("should fire on insertion", function (this: RealmContext, done) {
+      testObjects.addListener(
+        expectCollectionChangesOnEveryRun(done, [
+          { insertions: [], newModifications: [], oldModifications: [], deletions: [] },
+          { insertions: [0, 1], newModifications: [], oldModifications: [], deletions: [] },
+        ]),
+      );
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 1 });
+        this.realm.create("TestObject", { doubleCol: 2 });
+      });
+    });
+
+    it("should fire on update", function (this: RealmContext, done) {
+      testObjects.addListener(
+        expectCollectionChangesOnEveryRun(done, [
+          { insertions: [], newModifications: [], oldModifications: [], deletions: [] },
+          { insertions: [0, 1, 2], newModifications: [], oldModifications: [], deletions: [] },
+          { insertions: [], newModifications: [0, 2], oldModifications: [0, 2], deletions: [] },
```

Comment:
Would be good to see the difference between `newModifications` and `oldModifications`.  This could be done by doing a `delete` or `insert` combined with a modification in a single transaction.